File: LeetCode/1061/Solution.java
class Solution {
    public String smallestEquivalentString(String s1, String s2, String baseStr) {
        int[] parent = new int[26];
        for(int i=0;i<parent.length;++i) {
            parent[i] = i;
        }

        for(int i=0;i<s1.length();++i) {
            union(parent, s1.charAt(i), s2.charAt(i));
        }

        String result = "";
        for(int i=0;i<baseStr.length();++i) {
            char c = baseStr.charAt(i);
            result += (char)(find(parent, c)+'a');
        }
        return result;
    }

    private void union(int[] parent, char a, char b) {
        int pA = find(parent, a);
        int pB = find(parent, b);

        if(pA < pB) {
            parent[pB] = pA;
        } else {
            parent[pA] = pB;
        }
    }

    private int find(int[] parent, char a) {
        int index = a-'a';
        if(parent[index] == index) {
            return parent[index];
        }
        char c = (char)(parent[index]+'a');
        parent[index] = find(parent, c);
        return parent[index];
    }
}

File: LeetCode/1061/Solution.kt
class Solution {
    fun smallestEquivalentString(s1: String, s2: String, baseStr: String): String {
        val parents = mutableMapOf<Char, Char>()
        for(c in 'a'..'z') {
            parents[c] = c
        }

        for(i in 0 until s1.length) {
            union(parents, s1[i], s2[i])
        }

        var result = ""
        for(c in baseStr) {
            result += find(parents, c)
        }
        return result
    }

    fun find(parents: MutableMap<Char, Char>, node: Char): Char {
        if(parents[node] == node) {
            return parents[node]!!
        } else {
            parents[node] = find(parents, parents[node]!!)
        }
        return parents[node]!!
    }

    fun union(parents: MutableMap<Char, Char>, node1: Char, node2: Char) {
        val p1 = find(parents, node1)
        val p2 = find(parents, node2)

        if(p1 < p2) {
            parents[p2] = p1
        } else if(p1 > p2) {
            parents[p1] = p2
        }
    }
}
